netlogo-gui/src/main/editor/CodeCompletionPopup.scala

```diff
@@ -0,0 +1,135 @@
+            lastSuggested = suggestion
+            val token = compiler.getTokenAtPosition(editorArea.getText(), editorArea.getCaretPosition)
+            val position = editorArea.getCaretPosition
+            val doc = editorArea.getDocument.asInstanceOf[javax.swing.text.PlainDocument]
+            doc.replace(token.start, position - token.start, suggestion, null)
+            editorArea.setCaretPosition(token.start + suggestion.length)
+            isPopupEnabled = false
+          case _ =>
+            e.setSource(editorArea)
+            editorArea.dispatchEvent(e)
+        }
+      }
+
+      override def keyPressed(e: KeyEvent): Unit = {}
+
+      override def keyReleased(e: KeyEvent): Unit = {}
+    })
+  }
+
+  /**
+    * Makes the suggestion box initially and displays it.
+    * Should only be called when the user presses the shortcut to
+    * enable the suggestion. To update the suggestions list call showPopUp()
+    */
+  def enablePopup(): Unit = {
+    isPopupEnabled = true
+    showPopup()
+    val position = editorArea.getCaretPosition
+    window.setLocation(editorArea.getLocationOnScreen.x + editorArea.modelToView(position + 1).x,
+      editorArea.getLocationOnScreen.y + editorArea.modelToView(position).y + editorArea.getFont.getSize)
+  }
+
+  /**
+    * Updates the list of the suggestion box and makes it visible.
+    */
+  def showPopup(): Unit = {
+    if(isPopupEnabled) {
+      val token = compiler.getTokenAtPosition(editorArea.getText(), editorArea.getCaretPosition)
+      var list = Seq[String]()
+      // word only containd the current token till the cursor position
+      val word = if(token != null) token.text.substring(0, editorArea.getCaretPosition - token.start) else null
+      // popup not to be diplayed after user hits the enter key
+      if (word != null && !token.text.equals(lastSuggested)) {
+        list = autoSuggest.getSuggestions(word)
+        dlm.removeAllElements()
+        var i = 0
+        list.foreach(dlm.addElement(_))
+        if (!list.isEmpty) {
+          lastSuggested = ""
+          window.validate()
+          window.setVisible(true)
+          // Required to keep the caret in the editorArea visible
+          editorArea.getCaret.setVisible(true)
+        } else {
+          isPopupEnabled = false
+        }
+      }
+      if (list.isEmpty || token == null) {
+        window.setVisible(false)
+      }
+    }
+  }
+}
+
+/**
+  * This is the renderer to render elements of the list displayed
+  * in the suggestion box.
+  * @param editorArea
+  */
+class SuggestionListRenderer(editorArea: EditorArea) extends ListCellRenderer[String]{
+
+  override def getListCellRendererComponent(list: JList[_ <: String], value: String, index: Int, isSelected: Boolean, cellHasFocus: Boolean): Component = {
+    val label = new JLabel(value.asInstanceOf[String])
+    label.setForeground(if (DefaultTokenMapper.getCommand(value.asInstanceOf[String]).isEmpty) Color.BLUE
+    else new Color(0x551A8B))
```

Comment:
I would prefer these colors to come out of `window.SyntaxColors`, since those are the colors used by the EditorColorizer.